File: backend/src/main/java/org/sport/backend/entity/MatchResult.java
package org.sport.backend.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import org.sport.backend.dto.base.BaseEntity;
import org.sport.backend.constant.ResultStatus;

import java.util.*;

@Entity
@Table(name = "match_results")
@Getter
@Setter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
public class MatchResult extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID resultId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "match_id")
    private Match match;

    private UUID submitterId;

    @Column(name = "winning_team_number")
    private Integer winningTeamNumber;

    @ElementCollection
    @CollectionTable(
            name = "match_result_winners",
            joinColumns = @JoinColumn(name = "result_id")
    )
    @Column(name = "user_id")
    private List<UUID> winnerIds = new ArrayList<>();

    @ElementCollection
    @CollectionTable(
            name = "match_result_losers",
            joinColumns = @JoinColumn(name = "result_id")
    )
    @Column(name = "user_id")
    private List<UUID> loserIds = new ArrayList<>();

    @Enumerated(EnumType.STRING)
    private ResultStatus status;

    @ElementCollection
    @CollectionTable(
            name = "match_result_absent_users",
            joinColumns = @JoinColumn(name = "result_id")
    )
    @Column(name = "user_id")
    private List<UUID> absentUserIds = new ArrayList<>();


    @ElementCollection
    @CollectionTable(
            name = "match_result_rank_changes",
            joinColumns = @JoinColumn(name = "result_id")
    )
    @MapKeyColumn(name = "user_id")
    @Column(name = "point_change")
    private Map<UUID, Integer> rankChanges = new HashMap<>();
}

File: mobile/lib/models/match_result.dart
class MatchResultRequest {
  final String matchId;
  final int winningTeamNumber;
  final List<String>? absentUserIds;

  MatchResultRequest({
    required this.matchId,
    required this.winningTeamNumber,
    this.absentUserIds,
  });

  Map<String, dynamic> toJson() {
    return {
      'matchId': matchId,
      'winningTeamNumber': winningTeamNumber,
      'absentUserIds': absentUserIds ?? [],
    };
  }
}

class MatchResultResponse {
  final String resultId;
  final String matchId;
  final String submitterId;
  final int winningTeamNumber;
  final List<String> winnerIds;
  final List<String> loserIds;
  final String status;
  final List<String> absentUserIds;
  final Map<String, int>? rankChanges;

  MatchResultResponse({
    required this.resultId,
    required this.matchId,
    required this.submitterId,
    required this.winningTeamNumber,
    required this.winnerIds,
    required this.loserIds,
    required this.status,
    required this.absentUserIds,
    this.rankChanges,
  });

  factory MatchResultResponse.fromJson(Map<String, dynamic> json) {
    return MatchResultResponse(
      resultId: json['resultId']?.toString() ?? '',
      matchId: json['matchId']?.toString() ?? '',
      submitterId: json['submitterId']?.toString() ?? '',
      winningTeamNumber: int.tryParse(json['winningTeamNumber']?.toString() ?? '0') ?? 0,
      winnerIds: List<String>.from(json['winnerIds'] ?? []),
      loserIds: List<String>.from(json['loserIds'] ?? []),
      status: json['status']?.toString() ?? 'PENDING',
      absentUserIds: List<String>.from(json['absentUserIds'] ?? []),
      rankChanges: json['rankChanges'] != null ? Map<String, int>.from(json['rankChanges']) : null,
    );
  }
}